proxy/coco/src/request/waiting_room.rs

```diff
@@ -0,0 +1,303 @@
+pub enum Error {
+    #[error(transparent)]
+    Request(#[from] request::Error),
+    #[error("the URN '{0}' was not found in the waiting room")]
+    MissingUrn(RadUrn),
+    #[error("the state fetched from the waiting room was not the expected state")]
+    StateMismatch,
+}
+
+pub struct WaitingRoom<T> {
+    requests: HashMap<RadUrn, SomeRequest<T>>,
+}
+
+pub enum Strategy<R> {
+    First,
+    Newest,
+    Oldest,
+    Random(R),
+}
+
+impl<R> Strategy<R> {
+    pub fn next<'a, S: 'a, T: 'a + Clone + Ord>(
+        self,
+        mut requests: impl Iterator<Item = &'a Request<S, T>>,
+    ) -> Option<&'a Request<S, T>>
+    where
+        R: Rng,
+    {
+        match self {
+            Self::First => requests.next(),
+            Self::Newest => requests.max_by_key(|request| request.timestamp.clone()),
+            Self::Oldest => requests.min_by_key(|request| request.timestamp.clone()),
+            Self::Random(mut rng) => requests.choose(&mut rng),
+        }
+    }
+}
+
+// TODO(finto): Test scenario of "running" a request and updating the waiting room. Testing state
+// transitions.
+// TODO(finto): De/Serialize for waiting room.
+impl<T> WaitingRoom<T> {
+    pub fn new() -> Self {
+        Self {
+            requests: HashMap::new(),
+        }
+    }
+
+    pub fn create(&mut self, urn: RadUrn, timestamp: T) -> Option<SomeRequest<T>>
+    where
+        T: Clone,
+    {
+        match self.requests.get(&urn) {
+            None => {
+                let request = SomeRequest::Created(Request::new(urn.clone(), timestamp));
+                self.requests.insert(urn, request);
+                None
+            }
+            Some(request) => Some(request.clone()),
+        }
+    }
+
+    fn transition<Prev, Next>(
+        &mut self,
+        matcher: impl FnOnce(&SomeRequest<T>) -> Option<&Prev>,
+        transition: impl FnOnce(Prev) -> Result<Next, request::Error>,
+        urn: &RadUrn,
+    ) -> Result<Next, Error>
+    where
+        Prev: Clone,
+        Next: Into<SomeRequest<T>> + Clone,
+    {
+        match self.requests.get(urn) {
+            None => Err(Error::MissingUrn(urn.clone())),
+            Some(request) => match matcher(request) {
+                Some(previous) => {
+                    let next = transition(previous.clone())?;
+                    self.requests.insert(urn.clone(), next.clone().into());
+                    Ok(next)
+                }
+                None => Err(Error::StateMismatch),
+            },
+        }
+    }
+
+    pub fn requested(&mut self, urn: &RadUrn, timestamp: T) -> Result<Request<Requested, T>, Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Created(request) => Some(request),
+                _ => None,
+            },
+            |previous| Ok(previous.request(timestamp)),
+            urn,
+        )
+    }
+
+    pub fn attempted(
```

Comment:
As we refer to attempts both for queries and clones, this name might be bettah.

```suggestion
    pub fn queried(
```